source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,1896 @@
+actual value either does not exist or and matches the expected value. Matching
+the expected value should use the standard rules in `Evaluating Matches`_, which
+means that it may contain special operators.
+
+This operator is primarily used to assert driver-optional fields from the CRUD
+spec (e.g. ``insertedId`` for InsertOneResult, ``writeResult`` for
+BulkWriteException).
+
+An example of this operator used for a result's field follows::
+
+    expectedResult:
+      insertedId: { $$unsetOrMatches: 2 }
+
+An example of this operator used for an entire result follows::
+
+    expectedError:
+      expectedResult:
+        $$unsetOrMatches:
+          deletedCount: 0
+          insertedCount: 2
+          matchedCount: 0
+          modifiedCount: 0
+          upsertedCount: 0
+          upsertedIds: { }
+
+
+$$sessionLsid
+`````````````
+
+Syntax::
+
+    { $$sessionLsid: <string> }
+
+This operation is used for matching any value with the logical session ID of a
+`session entity <entity_session_>`_. The value will refer to a unique name of a
+session entity. The YAML file SHOULD use an `alias node`_ for a session entity's
+``id`` field (e.g. ``session: *session0``).
+
+An example of this operator follows::
+
+    command:
+      ping: 1
+      lsid: { $$sessionLsid: *session0 }
+
+
+Test Runner Implementation
+--------------------------
+
+The sections below describe instructions for instantiating the test runner,
+loading each test file, and executing each test within a test file. Test runners
+SHOULD NOT share state created by processing a test file with the processing of
+subsequent test files, and likewise for tests within a test file.
+
+
+Configuring the Test Runner
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The instructions in this section apply once for the test runner.
+
+The test runner MUST be configurable with a connection string (or equivalent
+configuration), which will be used to initialize the internal MongoClient and
+any `client entities <entity_client_>`_ (in combination with other URI options).
+This specification is not prescriptive about how this information is provided.
+For example, it may be read from an environment variable or configuration file.
+
+The test runner MAY determine the server version and topology type in advance to
+optimize any future `runOnRequirement`_ checks.
+
+The test runner SHOULD terminate any open transactions (see:
+`Terminating Open Transactions`_) before executing any tests.
+
+
+Executing a Test File
+~~~~~~~~~~~~~~~~~~~~~
+
+The instructions in this section apply for each test file loaded by the test
+runner. After processing a test file, test runners SHOULD reset any internal
+state that resulted from doing so. For example, an internal MongoClient created
+for one test file SHOULD NOT be shared with another.
+
+Test files, which may be YAML or JSON files, MUST be interpreted using an
+`Extended JSON`_ parser. The parser MUST accept relaxed and canonical Extended
```

Comment:
It is not clear to me how the runner is supposed to know whether to read the file as relaxed or canonical extended JSON. Can you clarify?

Reply:
Relaxed and canonical can be used in the same file (and are in various tests AFAIK). Per [Parsers](https://github.com/mongodb/specifications/blob/master/source/extended-json.rst#parsers):

> By default, a parser MUST accept values in either Canonical Extended JSON format or Relaxed Extended JSON format as described in this specification. A parser MAY allow users to restrict parsing to only Canonical Extended JSON format or only Relaxed Extended JSON format.

Would you like me to clarify that the parser should use its default mode and anticipate both?

Reply:
Will add a link to section of the Extended JSON spec for clarity.

Reply:
Link is now added. Resolving.